### PR TITLE
Use mol.GetBonds() instead of for loop

### PR DESCRIPTION
**Pull request by KnathanM**
This PR changes how we loop through bonds in `SimpleMoleculeMolGraphFeaturizer`. Previously we looped through the bonds manually as that was faster. That has been [fixed](https://github.com/rdkit/rdkit/pull/6419) and we can now use `.GetBonds()`. My timing experiments show that this version is only a little faster, but I think this version is more readable. 

## Timings
First I show that `.GetBonds()` is now fast
```
import chemprop
import rdkit
import torch
mol = rdkit.Chem.MolFromSmiles("CC1(C(N2C(S1)C(C2=O)NC(=O)CC3=CC=CC=C3)C(=O)O)C") # Penicillin
dp = chemprop.data.MoleculeDatapoint(mol)
ds = chemprop.data.MoleculeDataset([dp])

%timeit list(mol.GetBonds())
n_atoms = mol.GetNumAtoms()
%timeit [mol.GetBondBetweenAtoms(u, v) for u in range(n_atoms) for v in range(u+1, n_atoms)]

23.6 µs ± 696 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
69.1 µs ± 1.06 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```
My changes require getting the bond atom indexes so I also check that is not slow:
```
bond = mol.GetBondBetweenAtoms(0, 1)
%timeit u, v = bond.GetBeginAtomIdx(), bond.GetEndAtomIdx()

369 ns ± 6.79 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```
I test the featurization code by indexing a dataset which calls the featurizer. I also show the timing for this after caching for fun. I also show `edge_index` because the order of bonds is changed by this PR, but I believe that doesn't affect Chemprop as the bond order in `E` is also changed accordingly. 
```
# Original
%timeit ds[0]
ds.cache = True
%timeit ds[0]
ds[0][0].edge_index

340 µs ± 32.7 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
455 ns ± 6.35 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
array([[ 0,  1,  1,  2,  1,  5,  1, 22,  2,  3,  2, 19,  3,  4,  3,  7,
         4,  5,  4,  6,  6,  7,  6,  9,  7,  8,  9, 10, 10, 11, 10, 12,
        12, 13, 13, 14, 13, 18, 14, 15, 15, 16, 16, 17, 17, 18, 19, 20,
        19, 21],
       [ 1,  0,  2,  1,  5,  1, 22,  1,  3,  2, 19,  2,  4,  3,  7,  3,
         5,  4,  6,  4,  7,  6,  9,  6,  8,  7, 10,  9, 11, 10, 12, 10,
        13, 12, 14, 13, 18, 13, 15, 14, 16, 15, 17, 16, 18, 17, 20, 19,
        21, 19]])
```
Now for this PR
```
# GetBonds
%timeit ds[0]
ds.cache = True
%timeit ds[0]
ds[0][0].edge_index

263 µs ± 15.3 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
439 ns ± 8.86 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
array([[ 0,  1,  1,  2,  1,  5,  1, 22,  2,  3,  2, 19,  3,  4,  3,  7,
         4,  5,  4,  6,  6,  7,  6,  9,  7,  8,  9, 10, 10, 11, 10, 12,
        12, 13, 13, 14, 13, 18, 14, 15, 15, 16, 16, 17, 17, 18, 19, 20,
        19, 21],
       [ 1,  0,  2,  1,  5,  1, 22,  1,  3,  2, 19,  2,  4,  3,  7,  3,
         5,  4,  6,  4,  7,  6,  9,  6,  8,  7, 10,  9, 11, 10, 12, 10,
        13, 12, 14, 13, 18, 13, 15, 14, 16, 15, 17, 16, 18, 17, 20, 19,
        21, 19]])
```

## Related issues
We talk about this in #921. 

## Other ideas
I also tried using list comprehensions instead of a for loop, but it had the same timings as the for loop while also being less readable in my opinion. (I didn't bother checking this code for numerical accuracy as it didn't have better timings.)
```
if n_atoms == 0:
	V = np.zeros((1, self.atom_fdim), dtype=np.single)
else:
	V = np.array([self.atom_featurizer(a) for a in mol.GetAtoms()], dtype=np.single)

if atom_features_extra is not None:
	V = np.hstack((V, atom_features_extra))

x_e = np.array([self.bond_featurizer(bond) for bond in mol.GetBonds()], dtype=np.single)
x_e = np.hstack((x_e, bond_features_extra)) if bond_features_extra is not None else x_e
E = np.repeat(x_e, repeats=2, axis=0)

f1 = lambda bond: bond.GetBeginAtomIdx()
f2 = lambda bond: bond.GetEndAtomIdx()
edge_index = np.array([f(bond) for bond in mol.GetBonds() for f in (f1,f2)], int)
edge_index = np.vstack((edge_index, edge_index.reshape(-1, 2)[:, ::-1].flatten()))

rev_edge_index = np.arange(len(E)).reshape(-1, 2)[:, ::-1].ravel()

return MolGraph(V, E, edge_index, rev_edge_index)
```
```
# List comprehension
%timeit ds[0]
ds.cache = True
%timeit ds[0]

287 µs ± 6.53 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
439 ns ± 17.4 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

### CODE REVIEW

**Review by davidegraff — COMMENTED** [review]
they key check to perform that the two produce equivalent results when used as input to a message passing block, which I would believe that they do. Once that's checked (manually- no unit test), we can approve this and merge

**Comment by KnathanM** [comment]
Yep, the result of message passing is equivalent.

My test:
```
import chemprop
import rdkit
import torch
mol = rdkit.Chem.MolFromSmiles("CC1(C(N2C(S1)C(C2=O)NC(=O)CC3=CC=CC=C3)C(=O)O)C") # Penicillin
dp = chemprop.data.MoleculeDatapoint(mol)
ds = chemprop.data.MoleculeDataset([dp])
dl = chemprop.data.build_dataloader(ds, batch_size=1)
batch = next(iter(dl))
mg, *_ = batch
checkpoint_path =  "tests/data/example_model_v2_regression_mol.ckpt"
mpnn = chemprop.models.MPNN.load_from_checkpoint(checkpoint_path)
mpnn.message_passing(mg, None)
```

On `main` it gives:
```
tensor([[0.0146, 0.0000, 0.0711,  ..., 0.1050, 0.1969, 0.0000],
        [0.0836, 0.0000, 0.0000,  ..., 0.2042, 0.3485, 0.2322],
        [0.0000, 0.0000, 0.1735,  ..., 0.1548, 0.2860, 0.0000],
        ...,
        [0.0000, 0.0398, 0.2541,  ..., 0.0511, 0.0206, 0.0000],
        [0.0000, 0.0000, 0.0603,  ..., 0.0074, 0.1134, 0.0000],
        [0.0146, 0.0000, 0.0711,  ..., 0.1050, 0.1969, 0.0000]],
       grad_fn=<ReluBackward0>)
```

On `faster-bond-iteration` it gives:
```
tensor([[0.0146, 0.0000, 0.0711,  ..., 0.1050, 0.1969, 0.0000],
        [0.0836, 0.0000, 0.0000,  ..., 0.2042, 0.3485, 0.2322],
        [0.0000, 0.0000, 0.1735,  ..., 0.1548, 0.2860, 0.0000],
        ...,
        [0.0000, 0.0398, 0.2541,  ..., 0.0511, 0.0206, 0.0000],
        [0.0000, 0.0000, 0.0603,  ..., 0.0074, 0.1134, 0.0000],
        [0.0146, 0.0000, 0.0711,  ..., 0.1050, 0.1969, 0.0000]],
       grad_fn=<ReluBackward0>)
```

**Review by davidegraff — APPROVED** [review]
good catch on noticing that `Mol.getBonds()` speed was fixed!

**Review by shihchengli — APPROVED** [review]
